In [11]:
import pandas as pd
import numpy as np

In [12]:
%matplotlib inline
import matplotlib.pyplot as plt
import seaborn as sns

## Загрузка данных из репозитория

In [9]:
# train
train = pd.read_csv('https://github.com/Quantanomo/data_sets/raw/master/CAX_MortgageModeling_Train.csv')
train.head(2)


Unique_ID  MORTGAGE NUMBER  PROPERTY VALUE  MORTGAGE PAYMENT    GDS  \
0  CAX_Train_1                1          900000              5429  61.98   
1  CAX_Train_2                2          386000              2179  35.22   

     LTV    TDS  AMORTIZATION  MORTGAGE AMOUNT  RATE   ...      PROPERTY TYPE  \
0  65.00  71.63           360        1040000.0   4.0   ...    Single Detached   
1  74.29  40.65           360         390000.0   4.5   ...    Single Detached   

  TERM  FSA    AGE RANGE GENDER  INCOME INCOME TYPE  NAICS CODE  CREDIT SCORE  \
0   12  L4C     Under 25   Male  108000           8       44-45           681   
1   12  L9T  70 and over   Male   78000           2          56           710   

   RESULT  
0  FUNDED  
1  FUNDED  

[2 rows x 22 columns]

In [10]:
# test
test = pd.read_csv('https://github.com/Quantanomo/data_sets/raw/master/CAX_MortgageModeling_Test.csv')
test.head()

Unique_ID  MORTGAGE NUMBER  PROPERTY VALUE  MORTGAGE PAYMENT    GDS  \
0  CAX_Test_1            45643          430000              1861  34.51   
1  CAX_Test_2            45644          676000              1032  37.18   

     LTV   TDS  AMORTIZATION  MORTGAGE AMOUNT  RATE  ...    PROPERTY TYPE  \
0  75.00  35.6           360         322500.0  4.25  ...        High Rise   
1  55.51  37.5           360         375000.0  5.00  ...    Semi-Detached   

  TERM  FSA  AGE RANGE  GENDER INCOME INCOME TYPE  NAICS CODE  CREDIT SCORE  \
0   24  M9B      45-49    Male  76000           8          23           834   
1   12  M3N      55-59  Female  75000           2          56           804   

  RESULT  
0    NaN  
1    NaN  

[2 rows x 22 columns]

In [20]:
# Переименую колонки для удобного обращения к ним

train = train.rename(columns = {
    'MORTGAGE NUMBER': 'MORTGAGE_NUMBER',
    'PROPERTY VALUE': 'PROPERTY_VALUE',
    'MORTGAGE PAYMENT': 'MORTGAGE_PAYMENT',
    'MORTGAGE AMOUNT': 'MORTGAGE_AMOUNT',
    'PROPERTY TYPE': 'PROPERTY_TYPE',
    'AGE RANGE': 'AGE_RANGE',
    'INCOME TYPE': 'INCOME_TYPE',
    'NAICS CODE': 'NAICS_CODE',
    'CREDIT SCORE': 'CREDIT_SCORE'
})

train.head(2)

Unique_ID  MORTGAGE_NUMBER  PROPERTY_VALUE  MORTGAGE_PAYMENT    GDS  \
0  CAX_Train_1                1          900000              5429  61.98   
1  CAX_Train_2                2          386000              2179  35.22   

     LTV    TDS  AMORTIZATION  MORTGAGE_AMOUNT  RATE   ...      PROPERTY_TYPE  \
0  65.00  71.63           360        1040000.0   4.0   ...    Single Detached   
1  74.29  40.65           360         390000.0   4.5   ...    Single Detached   

  TERM  FSA    AGE_RANGE GENDER  INCOME INCOME_TYPE  NAICS_CODE  CREDIT_SCORE  \
0   12  L4C     Under 25   Male  108000           8       44-45           681   
1   12  L9T  70 and over   Male   78000           2          56           710   

   RESULT  
0  FUNDED  
1  FUNDED  

[2 rows x 22 columns]

In [21]:
test = test.rename(columns = {
    'MORTGAGE NUMBER': 'MORTGAGE_NUMBER',
    'PROPERTY VALUE': 'PROPERTY_VALUE',
    'MORTGAGE PAYMENT': 'MORTGAGE_PAYMENT',
    'MORTGAGE AMOUNT': 'MORTGAGE_AMOUNT',
    'PROPERTY TYPE': 'PROPERTY_TYPE',
    'AGE RANGE': 'AGE_RANGE',
    'INCOME TYPE': 'INCOME_TYPE',
    'NAICS CODE': 'NAICS_CODE',
    'CREDIT SCORE': 'CREDIT_SCORE'
})

test.head(2)

Unique_ID  MORTGAGE_NUMBER  PROPERTY_VALUE  MORTGAGE_PAYMENT    GDS  \
0  CAX_Test_1            45643          430000              1861  34.51   
1  CAX_Test_2            45644          676000              1032  37.18   

     LTV   TDS  AMORTIZATION  MORTGAGE_AMOUNT  RATE  ...    PROPERTY_TYPE  \
0  75.00  35.6           360         322500.0  4.25  ...        High Rise   
1  55.51  37.5           360         375000.0  5.00  ...    Semi-Detached   

  TERM  FSA  AGE_RANGE  GENDER INCOME INCOME_TYPE  NAICS_CODE  CREDIT_SCORE  \
0   24  M9B      45-49    Male  76000           8          23           834   
1   12  M3N      55-59  Female  75000           2          56           804   

  RESULT  
0    NaN  
1    NaN  

[2 rows x 22 columns]

### EDA

### Посмотрю на размерность данных

In [22]:
# Посмотрим на размер данных
train.shape, test.shape

((45642, 22), (15214, 22))

Вывод: фреймы одинаковы по своей структуре

### Проверю пропущенные значения и посмотрю на тип данных

In [23]:
# Функция принимает датафрейм и выдает кол-во и процент пропущенных значений и тип данных
def missing_data(data):
    total = data.isnull().sum()
    percent = (data.isnull().sum()/data.isnull().count()*100)
    tt = pd.concat([total, percent], axis=1, keys=['Total', 'Percent'])
    types = []
    for col in data.columns:
        dtype = str(data[col].dtype)
        types.append(dtype)
    tt['Types'] = types
    return(np.transpose(tt))

# train
missing_data(train)

Unique_ID MORTGAGE_NUMBER PROPERTY_VALUE MORTGAGE_PAYMENT      GDS  \
Total           0               0              0                0        0   
Percent         0               0              0                0        0   
Types      object           int64          int64            int64  float64   

             LTV      TDS AMORTIZATION MORTGAGE_AMOUNT     RATE   ...    \
Total          0        0            0               0        0   ...     
Percent        0        0            0               0        0   ...     
Types    float64  float64        int64         float64  float64   ...     

        PROPERTY_TYPE   TERM     FSA AGE_RANGE  GENDER INCOME INCOME_TYPE  \
Total               0      0       0         0       0      0           0   
Percent             0      0       0         0       0      0           0   
Types          object  int64  object    object  object  int64       int64   

        NAICS_CODE CREDIT_SCORE  RESULT  
Total            0            0       0  
Percent          0            0       0  
Types       object        int64  object  

[3 rows x 22 columns]

In [24]:
# test
missing_data(test)

Unique_ID MORTGAGE_NUMBER PROPERTY_VALUE MORTGAGE_PAYMENT      GDS  \
Total           0               0              0                0        0   
Percent         0               0              0                0        0   
Types      object           int64          int64            int64  float64   

             LTV      TDS AMORTIZATION MORTGAGE_AMOUNT     RATE   ...     \
Total          0        0            0               0        0   ...      
Percent        0        0            0               0        0   ...      
Types    float64  float64        int64         float64  float64   ...      

        PROPERTY_TYPE   TERM     FSA AGE_RANGE  GENDER INCOME INCOME_TYPE  \
Total               0      0       0         0       0      0           0   
Percent             0      0       0         0       0      0           0   
Types          object  int64  object    object  object  int64       int64   

        NAICS_CODE CREDIT_SCORE   RESULT  
Total            0            0    15214  
Percent          0            0      100  
Types       object        int64  float64  

[3 rows x 22 columns]

Вывод: Пропущенных значений нет (за исключением искомого значения в тестовом наборе). В наборах присутствуют количественные и номинативные признаки

### Посмотрю на характеристики количественных признаков

In [25]:
# train
train.describe()

MORTGAGE_NUMBER  PROPERTY_VALUE  MORTGAGE_PAYMENT           GDS  \
count      45642.00000    4.564200e+04      45642.000000  4.564200e+04   
mean       22821.50000    5.826110e+05       2392.148043  2.138033e+02   
std        13175.85483    4.382082e+05       1688.208721  3.263766e+04   
min            1.00000    1.400000e+04        141.000000 -3.704000e+01   
25%        11411.25000    3.400000e+05       1406.000000  1.984000e+01   
50%        22821.50000    4.770000e+05       2033.000000  2.921000e+01   
75%        34231.75000    6.850000e+05       2866.000000  3.587000e+01   
max        45642.00000    1.800000e+07      46479.000000  6.872300e+06   

                LTV           TDS  AMORTIZATION  MORTGAGE_AMOUNT  \
count  45642.000000  4.564200e+04  45642.000000     4.564200e+04   
mean      70.632932  4.125670e+02    355.309583     4.255704e+05   
std       11.566444  6.425859e+04     18.011798     3.020469e+05   
min        6.667000 -1.070000e+01     60.000000     1.500000e+04   
25%       65.000000  3.284000e+01    360.000000     2.459250e+05   
50%       75.000000  3.830000e+01    360.000000     3.575000e+05   
75%       80.000000  4.217000e+01    360.000000     5.120000e+05   
max       95.000000  1.346256e+07    480.000000     7.500000e+06   

               RATE          TERM        INCOME   INCOME_TYPE  CREDIT_SCORE  
count  45642.000000  45642.000000  4.564200e+04  45642.000000  45642.000000  
mean       4.651149     15.286578  1.251093e+05      5.315017    662.662811  
std        0.642371      7.443857  1.149974e+05      3.267761    132.818234  
min        2.500000      6.000000  1.000000e+03     -1.000000      0.000000  
25%        4.250000     12.000000  6.800000e+04      2.000000    623.000000  
50%        4.500000     12.000000  9.700000e+04      6.000000    677.000000  
75%        5.000000     12.000000  1.450000e+05      8.000000    736.000000  
max       10.000000     60.000000  3.860000e+06     20.000000    899.000000

In [26]:
# test
test.describe()

MORTGAGE_NUMBER  PROPERTY_VALUE  MORTGAGE_PAYMENT           GDS  \
count     15214.000000    1.521400e+04      15214.000000  1.521400e+04   
mean      53249.500000    5.790442e+05       2384.620350  2.306921e+03   
std        4392.047833    4.175277e+05       1608.625935  1.933853e+05   
min       45643.000000    3.600000e+04        208.000000 -4.878000e+01   
25%       49446.250000    3.370000e+05       1397.000000  1.997000e+01   
50%       53249.500000    4.740000e+05       2025.000000  2.930000e+01   
75%       57052.750000    6.800000e+05       2850.000000  3.610000e+01   
max       60856.000000    7.500000e+06      25032.000000  2.138046e+07   

                LTV           TDS  AMORTIZATION  MORTGAGE_AMOUNT  \
count  15214.000000  1.521400e+04  15214.000000     1.521400e+04   
mean      70.748041  2.754330e+03    355.208361     4.237224e+05   
std       11.580747  2.092715e+05     18.421023     2.898901e+05   
min        6.900000 -4.204000e+01    120.000000     3.338800e+02   
25%       65.000000  3.293000e+01    360.000000     2.441450e+05   
50%       75.000000  3.836000e+01    360.000000     3.562500e+05   
75%       80.000000  4.222000e+01    360.000000     5.111850e+05   
max       93.325000  2.274796e+07    480.000000     4.760000e+06   

               RATE          TERM        INCOME   INCOME_TYPE  CREDIT_SCORE  \
count  15214.000000  15214.000000  1.521400e+04  15214.000000   15214.00000   
mean       4.643010     15.405022  1.241874e+05      5.324438     663.11680   
std        0.633511      7.531778  1.110122e+05      3.303275     133.94336   
min        2.500000      6.000000  1.000000e+03     -1.000000       0.00000   
25%        4.250000     12.000000  6.700000e+04      2.000000     624.00000   
50%        4.500000     12.000000  9.700000e+04      6.000000     678.00000   
75%        5.000000     12.000000  1.440000e+05      8.000000     736.00000   
max       10.000000     60.000000  2.125000e+06     19.000000     899.00000   

       RESULT  
count     0.0  
mean      NaN  
std       NaN  
min       NaN  
25%       NaN  
50%       NaN  
75%       NaN  
max       NaN

MORTGAGE NUMBER (Номер ипотеки) - ни о чем нам не говорит. Судя по номерам, я исследую 2/3 (train) и 1/3 (test) одного датафрейма (номера ипотек с 1 по 60856)

PROPERTY VALUE (Стоимость имущества) - Среднее, медиана и sd в наборах примерно одинаковы

MORTGAGE PAYMENT (Выплата по ипотеке) - Среднее, медиана и sd в наборах примерно одинаковы

GDS (Годовые расходы на жилье / Годовой доход) - Медианы одинаковы, средние и sd различны

LTV (сумма ипотеки / оценочная стоимость недвижимости) - Среднее, медиана и sd в наборах примерно одинаковы

TDS (Все годовые расходы / Годовой доход) - Схоже по параметрам с GDS

AMORTIZATION (Амортизация в месяцах) - Амортизация в наборах одинакова

MORTGAGE AMOUNT	(Сумма ипотеки в $) - Среднее, медиана и sd в наборах примерно одинаковы

RATE (Процентная ставка) - Ставки в наборах равны

TERM (Срок ипотеки) - Сроки ипотеки одинаковы

INCOME (Доходы) - Доходы одинаковы

INCOME TYPE (Тип дохода) - Это статусы занятости (номинативная переменная)

CREDIT SCORE (Рейтинг кредитного бюро) - рейтинги в наборах похожи

----------- НОМИНАТИВНЫЕ ПРИЗНАКИ

PROPERTY TYPE (Тип недвижимости)

FSA (Почтовый индекс)

AGE RANGE (Возрастная категория)

GENDER (Пол)

NAICS CODE (Категория работы)

https://www.kaggle.com/gpreda/santander-eda-and-prediction